### PR TITLE
Unit tests for datatype special characters

### PR DESCRIPTION
**Pull request by ShivanshGahlot**
### Does your PR have changes to on-disk structures that can cause upgrade issues? 

---

#### On-disk structures potentially causing breaking changes:
| Component        
| :----------------------------------------------: | 
| MetaDB                                           |
| Name registry json                               |
| Data File Descriptor Json                        |
| Export Snapshot Status Json                      |
| Callhome Json                                    |
| Export Status Json                               |
| YugabyteD Tables                                 |
| TargetDB Metadata Tables                         |
| Schema Dump                                      |
| AssessmentDB                                     |
| Migration Assessment Report Json                 |
| Import Data State                                |
| Export and import data queue                     |
| Data .sql files of tables                        |

### CODE REVIEW

File: yb-voyager/cmd/live_migration_integration_test.go [hunk 1]
```diff
@@ -1686,3 +1686,1332 @@ $$ LANGUAGE plpgsql;`,
...
+		`test_schema."uuid_ltree_edge_cases"`:     3, // 3 rows with UUID/LTREE edge cases
+		`test_schema."map_edge_cases"`:            3, // 3 rows with MAP edge cases
+		`test_schema."interval_edge_cases"`:       3, // 3 rows with INTERVAL edge cases
+		`test_schema."zonedtimestamp_edge_cases"`: 3, // 3 rows with ZONEDTIMESTAMP edge cases
+		`test_schema."decimal_edge_cases"`:        3, // 3 rows with DECIMAL edge cases
+	}, 120) // 1 minute should be plenty for small rows
+	testutils.FatalIfError(t, err, "failed to wait for snapshot complete")
+
+	t.Log("=== Validating snapshot data ===")
+	err = lm.ValidateDataConsistency([]string{`test_schema."string_edge_cases"`, `test_schema."json_edge_cases"`, `test_schema."enum_edge_cases"`, `test_schema."bytes_edge_cases"`, `test_schema."datetime_edge_cases"`, `test_schema."uuid_ltree_edge_cases"`, `test_schema."map_edge_cases"`, `test_schema."interval_edge_cases"`, `test_schema."zonedtimestamp_edge_cases"`, `test_schema."decimal_edge_cases"`}, "id")
```
Comment (makalaaneesh) [inline yb-voyager/cmd/live_migration_integration_test.go]: I'd recommend doing a manual verification once (like actually do a select * once and manually verify the outputs ) just because this is some function that we have written (so it could potentially be buggy)

File: yb-voyager/src/tgtdb/suites/yugabytedbSuite_test.go [hunk 2]
```diff
@@ -148,3 +148,1523 @@ func TestBytesConversionWithFormatting(t *testing.T) {
...
+
+		t.Logf("✓ 200MB string: %d bytes → %d bytes in %v", size, len(result), duration)
+		t.Logf("✓ Throughput: %.2f MB/s", float64(size)/(1024*1024)/duration.Seconds())
+	})
+}
+
+// Test 2.1: TestJsonConversionBasic
+// Tests JSON/JSONB type with basic valid JSON (no single quotes inside JSON strings)
+// NOTE: The converter uses quoteValueIfRequiredWithEscaping which doubles single quotes,
+// but this creates INVALID JSON. Therefore, we only test JSON that doesn't contain
```
Comment (makalaaneesh) [inline yb-voyager/src/tgtdb/suites/yugabytedbSuite_test.go]: why does this matter? we just need to assert it properly, why does it matter if it's an invalid json or not

File: yb-voyager/cmd/live_migration_integration_test.go [hunk 3]
```diff
@@ -1686,3 +1686,1332 @@ $$ LANGUAGE plpgsql;`,
 	testutils.FatalIfError(t, err, "failed to wait for cutover complete")
 
 }
+
+// TestLiveMigrationWithDatatypeEdgeCases tests live migration with various datatypes
```
[the remark this answers is not in the document]
Comment (makalaaneesh) [reply yb-voyager/cmd/live_migration_integration_test.go]: some of these might be NA, but let's aim for full coverage, and also document the coverage matrix at the beginning of the test @ShivanshGahlot

File: yb-voyager/cmd/live_migration_integration_test.go [hunk 4]
```diff
@@ -1686,3 +1686,1332 @@ $$ LANGUAGE plpgsql;`,
...
+// containing special characters and edge cases that require proper escaping.
+// Currently testing: STRING datatype with backslashes, quotes, newlines, tabs, Unicode, etc.
+// This test verifies that the datatype converter properly handles edge cases during CDC streaming.
+// Aligned with unit tests in yugabytedbSuite_test.go
+func TestLiveMigrationWithDatatypeEdgeCases(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ShivanshGahlot) [reply yb-voyager/cmd/live_migration_integration_test.go]: Added these

File: yb-voyager/cmd/live_migration_integration_test.go [hunk 5]
```diff
@@ -1686,3 +1686,3648 @@ $$ LANGUAGE plpgsql;`,
...
+				text_windows_path,
+				text_sql_injection,
+				text_unicode,
+				text_empty,
+				text_null_string
+			) VALUES
+			(
+				'path\to\file',                          -- literal backslash-t, backslash-o
+				'It''s a test',                          -- single quote (SQL escaped)
+				'line1' || E'\u2028' || 'line2',         -- TODO 6: Unicode line separator (U+2028)
```
Comment (makalaaneesh) [inline yb-voyager/cmd/live_migration_integration_test.go]: pls remove the "TODO" comments everywhere. It just ultimately adds noise

File: yb-voyager/cmd/live_migration_integration_test.go [hunk 6]
```diff
@@ -1686,3 +1686,3648 @@ $$ LANGUAGE plpgsql;`,
...
+			`UPDATE test_schema.string_edge_cases
+		SET text_with_quote = NULL,
+		    text_empty = NULL
+		WHERE id = 5;`,
+
+			// STRING: NULL restore
+			`UPDATE test_schema.string_edge_cases
+		SET text_with_quote = 'restored from NULL',
+		    text_empty = 'restored'
+		WHERE id = 5 AND text_with_quote IS NULL;`,
```
Comment (makalaaneesh) [inline yb-voyager/cmd/live_migration_integration_test.go]: If we restore them, how do we validate that they are properly set to NULL?

Same for all the cases.

File: yb-voyager/cmd/live_migration_integration_test.go [hunk 7]
```diff
@@ -1686,3 +1686,3648 @@ $$ LANGUAGE plpgsql;`,
...
+//   - UPDATEs:  62 (includes NULL transitions + Unicode separators + literal \n/actual newline + single quotes in JSON)
+//   - DELETEs:  12 (1 per datatype, targets snapshot row 4)
+//
+// NULL Transition Coverage:
+//   - 11 datatypes with full NULL↔non-NULL↔NULL transitions (both directions)
+//   - Tested in both forward and fallback streaming
+//   - Covers row 5 (non-NULL→NULL→non-NULL) for STRING, JSON, DECIMAL
+//   - Covers row 6 (NULL→non-NULL→NULL) for all 11 datatypes
+
+func getDatatypeEdgeCasesTestConfig() *TestConfig {
```
Comment (makalaaneesh) [inline yb-voyager/cmd/live_migration_integration_test.go]: let's move this test to a separate file altogether. It's too huge

Comment (ShivanshGahlot) [reply yb-voyager/cmd/live_migration_integration_test.go]: Done

File: yb-voyager/cmd/live_migration_integration_test.go [hunk 8]
```diff
@@ -1686,3 +1686,3648 @@ $$ LANGUAGE plpgsql;`,
...
+			// BOOLEAN Row 3: TRUE/NULL - will be deleted in forward
+			`INSERT INTO test_schema.boolean_edge_cases (bool_simple, bool_nullable) VALUES (TRUE, NULL);`,
+			// BOOLEAN Row 4: FALSE/NULL - will be deleted in fallback
+			`INSERT INTO test_schema.boolean_edge_cases (bool_simple, bool_nullable) VALUES (FALSE, NULL);`,
+			// BOOLEAN Row 5: NULL/TRUE
+			`INSERT INTO test_schema.boolean_edge_cases (bool_simple, bool_nullable) VALUES (NULL, TRUE);`,
+			// BOOLEAN Row 6: NULL/NULL - for NULL transitions
+			`INSERT INTO test_schema.boolean_edge_cases (bool_simple, bool_nullable) VALUES (NULL, NULL);`,
+		},
+		SourceDeltaSQL: []string{
```
[the remark this answers is not in the document]
Comment (ShivanshGahlot) [reply yb-voyager/cmd/live_migration_integration_test.go]: Grouped statements belonging to the same table together.

File: yb-voyager/cmd/live_migration_integration_test.go (hunk 6 above)
Comment (ShivanshGahlot) [reply yb-voyager/cmd/live_migration_integration_test.go]: Ahh this got messed up earlier. I have fixed it in a newer commit.
So this is the correct pattern

row x: NULL -> not null (forward direction) -> NULL (fall back)
row y: not null -> NULL (forward direction) -> non null (fall back)

File: yb-voyager/cmd/live_migration_integration_test.go [hunk 9]
```diff
@@ -2630,3 +2630,621 @@ VALUES (INTERVAL '7 years', INTERVAL '120 days', INTERVAL '15 hours', INTERVAL '
 	testutils.FatalIfError(t, err, "failed to wait for cutover source complete")
 
 }
+
+// TestLiveMigrationWithDatatypeEdgeCases tests live migration with various datatypes
```
Comment (makalaaneesh) [inline yb-voyager/cmd/live_migration_integration_test.go]: let's move all the tests you wrote to a new file, not just the helper
